File: android-agent/app/src/main/java/com/androidscript/agent/runtime/Token.kt
package com.androidscript.agent.runtime

/**
 * Token types for AndroidScript lexer
 */
enum class TokenType {
    // Literals
    IDENTIFIER,
    STRING,
    INTEGER,
    FLOAT,
    TRUE,
    FALSE,
    NULLPTR,

    // Operators
    PLUS,           // +
    MINUS,          // -
    MULTIPLY,       // *
    DIVIDE,         // /
    MODULO,         // %
    ASSIGN,         // =
    EQUAL,          // ==
    NOT_EQUAL,      // !=
    LESS,           // <
    LESS_EQUAL,     // <=
    GREATER,        // >
    GREATER_EQUAL,  // >=
    LOGICAL_AND,    // &&
    LOGICAL_OR,     // ||
    LOGICAL_NOT,    // !

    // Delimiters
    LPAREN,         // (
    RPAREN,         // )
    LBRACE,         // {
    RBRACE,         // }
    LBRACKET,       // [
    RBRACKET,       // ]
    COMMA,          // ,
    DOT,            // .
    COLON,          // :
    SEMICOLON,      // ;

    // Keywords
    IF,
    ELSE,
    WHILE,
    FOR,
    FOREACH,
    REPEAT,
    UNTIL,
    FUNCTION,
    RETURN,
    BREAK,
    CONTINUE,
    TRY,
    CATCH,
    FINALLY,
    IN,

    // Directives
    DIRECTIVE,      // #include, #import, etc.

    // Special
    NEWLINE,
    END_OF_FILE,
    INVALID
}

/**
 * Token data class representing a lexical token
 */
data class Token(
    val type: TokenType,
    val lexeme: String,
    val line: Int,
    val column: Int,
    val intValue: Long = 0,
    val floatValue: Double = 0.0,
    val boolValue: Boolean = false
) {
    companion object {
        fun invalid() = Token(TokenType.INVALID, "", 0, 0)
        fun eof(line: Int, column: Int) = Token(TokenType.END_OF_FILE, "", line, column)
    }
}

/**
 * Keyword mapping for identifier lookup
 */
object Keywords {
    val map = mapOf(
        "if" to TokenType.IF,
        "else" to TokenType.ELSE,
        "while" to TokenType.WHILE,
        "for" to TokenType.FOR,
        "foreach" to TokenType.FOREACH,
        "repeat" to TokenType.REPEAT,
        "until" to TokenType.UNTIL,
        "function" to TokenType.FUNCTION,
        "return" to TokenType.RETURN,
        "break" to TokenType.BREAK,
        "continue" to TokenType.CONTINUE,
        "try" to TokenType.TRY,
        "catch" to TokenType.CATCH,
        "finally" to TokenType.FINALLY,
        "in" to TokenType.IN,
        "true" to TokenType.TRUE,
        "false" to TokenType.FALSE,
        "null" to TokenType.NULLPTR
    )

    fun lookup(identifier: String): TokenType {
        return map[identifier.lowercase()] ?: TokenType.IDENTIFIER
    }
}

File: ios-agent/iOSAgent/Runtime/Token.swift
import Foundation

/**
 * Token types for AndroidScript
 */
enum TokenType {
    // Literals
    case identifier
    case string
    case integer
    case float
    case `true`
    case `false`
    case nullptr

    // Operators
    case plus, minus, multiply, divide, modulo
    case assign
    case equal, notEqual
    case less, lessEqual, greater, greaterEqual
    case logicalAnd, logicalOr, logicalNot

    // Delimiters
    case lparen, rparen
    case lbrace, rbrace
    case lbracket, rbracket
    case comma, dot, semicolon

    // Keywords
    case `if`, `else`
    case `while`, `for`, foreach, `repeat`, until
    case function, `return`
    case `break`, `continue`
    case `try`, `catch`, finally
    case `in`
    case `var`, `let`, const

    // Special
    case newline
    case endOfFile
}

/**
 * Token structure
 */
struct Token {
    let type: TokenType
    let lexeme: String
    let line: Int
    let column: Int

    // Value storage for literals
    let intValue: Int64
    let floatValue: Double
    let boolValue: Bool

    init(type: TokenType, lexeme: String, line: Int, column: Int,
         intValue: Int64 = 0, floatValue: Double = 0.0, boolValue: Bool = false) {
        self.type = type
        self.lexeme = lexeme
        self.line = line
        self.column = column
        self.intValue = intValue
        self.floatValue = floatValue
        self.boolValue = boolValue
    }

    static func eof(line: Int, column: Int) -> Token {
        return Token(type: .endOfFile, lexeme: "", line: line, column: column)
    }

    static func identifier(_ name: String, line: Int, column: Int) -> Token {
        return Token(type: .identifier, lexeme: name, line: line, column: column)
    }

    static func string(_ value: String, line: Int, column: Int) -> Token {
        return Token(type: .string, lexeme: value, line: line, column: column)
    }

    static func integer(_ value: Int64, line: Int, column: Int) -> Token {
        return Token(type: .integer, lexeme: String(value), line: line, column: column, intValue: value)
    }

    static func float(_ value: Double, line: Int, column: Int) -> Token {
        return Token(type: .float, lexeme: String(value), line: line, column: column, floatValue: value)
    }
}

/**
 * Keyword lookup table
 */
struct Keywords {
    private static let map: [String: TokenType] = [
        "if": .if,
        "else": .else,
        "while": .while,
        "for": .for,
        "foreach": .foreach,
        "repeat": .repeat,
        "until": .until,
        "function": .function,
        "return": .return,
        "break": .break,
        "continue": .continue,
        "try": .try,
        "catch": .catch,
        "finally": .finally,
        "in": .in,
        "var": .var,
        "let": .let,
        "const": .const,
        "true": .true,
        "false": .false,
        "null": .nullptr,
        "nil": .nullptr
    ]

    static func lookup(_ identifier: String) -> TokenType {
        return map[identifier] ?? .identifier
    }

    static func isKeyword(_ identifier: String) -> Bool {
        return map[identifier] != nil
    }
}
